### PR TITLE
New Dataset Cycle_GAN added

### PR DESCRIPTION
**Pull request by ChanchalKumarMaji**
Added Cycle_GAN Dataset.
Dataset Request #24 

@rsepassi @cyfra @Conchylicultor @dynamicwebpaige @tfds-copybara , Please review.

### CODE REVIEW

File: tensorflow_datasets/image/__init__.py [hunk 1]
```diff
@@ -38,3 +38,5 @@
 from tensorflow_datasets.image.quickdraw import QuickdrawBitmap
 from tensorflow_datasets.image.rock_paper_scissors import RockPaperScissors
 from tensorflow_datasets.image.svhn import SvhnCropped
+from tensorflow_datasets.image.cycle_gan import CycleGANConfig
```
Comment (cyfra) [inline tensorflow_datasets/image/__init__.py]: this line is probably not needed.

File: tensorflow_datasets/image/cycle_gan.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      raise ValueError("data must be one of %s" % _DATA_OPTIONS)
+  
+    super(CycleGANConfig, self).__init__(**kwargs)
+    self.data = data
+
+    
+class CycleGAN(tfds.core.GeneratorBasedBuilder):
+  """CycleGAN dataset."""
+  
+  VERSION = tfds.core.Version("2.0.0")
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: what's the reason for 2.0?

File: tensorflow_datasets/image/cycle_gan.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+  
+  BUILDER_CONFIGS = [
+      CycleGANConfig(
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: this is a lot of copy-paste (which is prone to typos).

Consider doing:

BUILDER_CONFIGS = [
  CycleGANConfig(name=config_name, ...) for config_name in _DATA_OPTIONS]

File: tensorflow_datasets/image/cycle_gan.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+from __future__ import print_function
+
+import os
+
+import numpy as np
+import tensorflow as tf
+
+from tensorflow_datasets.core import api_utils
+import tensorflow_datasets.public_api as tfds
+
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: citation missing

File: tensorflow_datasets/image/cycle_gan.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+_DL_URL = "https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/"
+_DL_URLS = {
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: if the names are matching the zip files, you can do:

_DL_URLS = { name: _DL_URL + name + ".zip" for name in _DATA_OPTIONS}

File: tensorflow_datasets/image/cycle_gan.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            "label": tfds.features.ClassLabel(names=["A", "B"]), 
+        }),
+        supervised_keys=("image", "label"),
+        urls=["https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/"],
+    )
+
+  def _split_generators(self, dl_manager):
+    url = _DL_URLS[self.builder_config.name]
+    data_dirs = dl_manager.download_and_extract(url) 
+    print(data_dirs)
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: remove

File: tensorflow_datasets/image/cycle_gan.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    path_to_dataset = data_dirs + "/" + tf.io.gfile.listdir(data_dirs)[0]
+    print(path_to_dataset)
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: remove

File: tensorflow_datasets/image/cycle_gan.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        }),
+        supervised_keys=("image", "label"),
+        urls=["https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/"],
+    )
+
+  def _split_generators(self, dl_manager):
+    url = _DL_URLS[self.builder_config.name]
+    data_dirs = dl_manager.download_and_extract(url) 
+    print(data_dirs)
+    path_to_dataset = data_dirs + "/" + tf.io.gfile.listdir(data_dirs)[0]
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: use os.path.join

File: tensorflow_datasets/image/cycle_gan.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            name="testB",
+            num_shards=1,
+            gen_kwargs={"files": testB_files,
+                        "label": "B",
+            }
+        ),
+    ]
+
+  def _generate_examples(self, files, label): 
+    path=files[0]+"/"
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: os.path.join

File: tensorflow_datasets/image/cycle_gan.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    images=tf.io.gfile.listdir(path)  
+
+    for image in images:
+      yield{
+          "image": path+"/"+image,
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan.py]: os.path.join

File: tensorflow_datasets/image/cycle_gan_test.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+      "trainB": 2,
+      "testB": 2,
+  }
+  
+if __name__ == "__main__":
+  testing.test_main()
+  
+  
+  
+
```
Comment (cyfra) [inline tensorflow_datasets/image/cycle_gan_test.py]: remove lines from the end of the file.

File: tensorflow_datasets/image/cycle_gan.py (hunk 3 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: One thing I want to mention here is that, the dataset <code>ae_photos</code> has only testA and testB, (trainA and trainB missing). Should I remove such sets ?

File: tensorflow_datasets/image/__init__.py (hunk 1 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/__init__.py]: Ok, removing this line produces no error, but can you please tell me why it is used [here](https://github.com/tensorflow/datasets/blob/96fe8c3156b43eb776456522ac739a0a6cdb86d2/tensorflow_datasets/text/__init__.py#L19), [here](https://github.com/tensorflow/datasets/blob/96fe8c3156b43eb776456522ac739a0a6cdb86d2/tensorflow_datasets/text/__init__.py#L21), [here](https://github.com/tensorflow/datasets/blob/96fe8c3156b43eb776456522ac739a0a6cdb86d2/tensorflow_datasets/audio/__init__.py#L19) and [here](https://github.com/tensorflow/datasets/blob/96fe8c3156b43eb776456522ac739a0a6cdb86d2/tensorflow_datasets/video/__init__.py#L21). Should I modify these also ?

File: tensorflow_datasets/image/cycle_gan.py (hunk 2 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: What should I use here ? Should I use <code>tfds.core.Version('0.1.0')</code>. Please suggest. I see something as <code>MAJOR.MINOR.PATCH</code>.

File: tensorflow_datasets/image/cycle_gan.py (hunk 4 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: I cannot find any CITATION [here](https://people.eecs.berkeley.edu/~taesung_park/CycleGAN/datasets/). What to do ?

File: tensorflow_datasets/image/cycle_gan.py (hunk 6 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: Done

File: tensorflow_datasets/image/cycle_gan.py (hunk 7 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: Done

**Comment by ChanchalKumarMaji** [comment]
@cyfra I have made the necessary changes. Please review. Thanks.

File: tensorflow_datasets/image/__init__.py (hunk 1 above)
Comment (cyfra) [reply tensorflow_datasets/image/__init__.py]: you're right. My bad.

File: tensorflow_datasets/image/cycle_gan.py (hunk 4 above)
Comment (cyfra) [reply tensorflow_datasets/image/cycle_gan.py]: https://arxiv.org/abs/1703.10593

File: tensorflow_datasets/image/cycle_gan.py (hunk 2 above)
Comment (cyfra) [reply tensorflow_datasets/image/cycle_gan.py]: We usually start with 0.1.0 - if dataset is under development and then eventually 1.0.0

Using 2.0 is not an error per se - but it suggests that we had an earlier version

File: tensorflow_datasets/image/cycle_gan.py (hunk 3 above)
Comment (cyfra) [reply tensorflow_datasets/image/cycle_gan.py]: Do you know if it is missing on purpose ?

In such case, I'd leave it out (And leave a comment saying why ae_photos is missing)
If it ever becomes needed, we can always add it back.

File: tensorflow_datasets/image/cycle_gan.py (hunk 2 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: Yes, I changed it to "0.1.0". Thanks.

File: tensorflow_datasets/image/cycle_gan.py (hunk 3 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: I don't know why they are missing, but I am removing them and adding a comment that these datasets are not present.

File: tensorflow_datasets/image/cycle_gan.py (hunk 4 above)
Comment (ChanchalKumarMaji) [reply tensorflow_datasets/image/cycle_gan.py]: Done. Please Check.

**Comment by ChanchalKumarMaji** [comment]
What is it that the "kokoro" is failing ? Any suggestions.

**Comment by ChanchalKumarMaji** [comment]
I think this dataset is ready now. Please Check. Thanks.

**Comment by cyfra** [comment]
Thanks - I'll try to run the final set of tests and merge it.

**Comment by cyfra** [comment]
@ChanchalKumarMaji - please generate fake images instead of including real ones.

**Comment by ChanchalKumarMaji** [comment]
@cyfra , I removed the real images and replaced them with fake_examples. Please check. Thanks.

**Comment by cyfra** [comment]
Great, thanks.

**Comment by ChanchalKumarMaji** [comment]
@cyfra @rsepassi , why is the <code>kokoro</code> build failing. I cannot understand. Thanks.

**Comment by cyfra** [comment]
Kokoro was failing with:
E   AssertionError: Splits 'testA' and 'trainA' are overlapping. Are you sure you want to have the same objects in those splits? If yes, add one one of them to OVERLAPPING_SPLITS class attribute.

you can see this error, when you run the test locally too.

Anyway - I've already fixed it in the local version that is being merged now, so you don't have to do any additional changes.

**Comment by ChanchalKumarMaji** [comment]
My bad, I should not have used the same samples in "trainA" and "testA". Actually, I didn't ran the test locally after changing the images. If any changes are required then please do tell me.

**Comment by ChanchalKumarMaji** [comment]
Ok, I have used different samples in the splits, and I ran the test locally with no errors. Thanks.

**Comment by cyfra** [comment]
Merged. Thanks @ChanchalKumarMaji

**Pull request closed without merging**